File: boj/Stack/BOJ_2493.swift
// 백준 2493 탑
// 21.05.02

import Foundation

let n = Int(readLine()!)!
var tops = readLine()!.split(separator: " ").map { Int(String($0))! }
var result = [Int](repeating: 0, count: n)
var stack = [Int]()

for i in stride(from: n - 1, through: 0, by: -1) {
  if stack.isEmpty {
    stack.append(i)
  } else {
    while !stack.isEmpty && tops[stack.last!] < tops[i] {
      let lowTopIndex = stack.popLast()!
      result[lowTopIndex] = i + 1
    }
    stack.append(i)
  }
}

for res in result {
  print(res, terminator: " ")
}

File: boj/Stack/BOJ_2493.py
# 백준 2493 탑
# 21.05.02

import sys
import heapq
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
# sys.setrecursionlimit(10**6)
rd = sys.stdin.readline
INF = int(1e9)

n = int(rd())
tops = list(map(int, rd().split()))
result = [0] * n
stack = []

for i in range(n-1, -1, -1):
    while stack and tops[stack[-1]] < tops[i]:
        result[stack[-1]] = i + 1
        stack.pop()
    stack.append(i)

for res in result:
    print(res, end=" ")